File: tasks/money-rollup/python/src/fx.py
"""Currency rate lookup."""

from money import parse_rate


def build_graph(currencies, rates):
    edges = {}
    for rate in rates:
        edges[(rate["from"], rate["to"])] = parse_rate(rate["rate"])
    return edges


def factor(edges, source, target):
    if source == target:
        return 1.0
    if (source, target) not in edges:
        raise ValueError("no conversion rate")
    return edges[(source, target)]

File: tasks/money-rollup/typescript/src/fx.ts
import { parseRate } from "./money";

export interface Rate {
  from: string;
  to: string;
  rate: string;
}

export type Graph = Map<string, number>;

function edgeKey(source: string, target: string): string {
  return source + " " + target;
}

export function buildGraph(
  currencies: Record<string, number>,
  rates: readonly Rate[],
): Graph {
  const edges: Graph = new Map();
  for (const rate of rates) {
    edges.set(edgeKey(rate.from, rate.to), parseRate(rate.rate));
  }
  return edges;
}

export function factor(edges: Graph, source: string, target: string): number {
  if (source === target) {
    return 1;
  }
  const direct = edges.get(edgeKey(source, target));
  if (direct === undefined) {
    throw new Error("no conversion rate");
  }
  return direct;
}
